x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/ChainingInputStream.java

```diff
@@ -0,0 +1,214 @@
+
+    /**
+     * value for the current input stream when there are no subsequent streams left, i.e. when
+     * {@link #nextComponent(InputStream)} returns {@code null}
+     */
+    protected static final InputStream EXHAUSTED_MARKER = InputStream.nullInputStream(); // protected for tests
+
+    /**
+     * The instance of the currently in use component input stream,
+     * i.e. the instance currently servicing the read and skip calls on the {@code ChainingInputStream}
+     */
+    protected InputStream currentIn; // protected for tests
+    /**
+     * The instance of the component input stream at the time of the last {@code mark} call.
+     */
+    protected InputStream markIn; // protected for tests
+    /**
+     * {@code true} if {@link #close()} has been called; any subsequent {@code read}, {@code skip}
+     * {@code available} and {@code reset} calls will throw {@code IOException}s
+     */
+    private boolean closed;
+
+    /**
+     * This method is responsible for generating the component input streams.
+     * It is passed the current input stream and must return the successive one,
+     * or {@code null} if the current component is the last one. It is passed the {@code null} value
+     * at the very start, when no component input stream has yet been obtained.
+     */
+    abstract @Nullable InputStream nextComponent(@Nullable InputStream currentComponentIn) throws IOException;
+
+    @Override
+    public int read() throws IOException {
+        ensureOpen();
+        do {
+            int byteVal = currentIn == null ? -1 : currentIn.read();
+            if (byteVal != -1) {
+                return byteVal;
+            }
+        } while (nextIn());
+        return -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        Objects.checkFromIndexSize(off, len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        do {
+            int bytesRead = currentIn == null ? -1 : currentIn.read(b, off, len);
+            if (bytesRead != -1) {
+                return bytesRead;
+            }
+        } while (nextIn());
+        return -1;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        ensureOpen();
+        if (n <= 0) {
+            return 0;
+        }
+        if (currentIn == null) {
+            nextIn();
+        }
+        long bytesRemaining = n;
+        while (bytesRemaining > 0) {
+            long bytesSkipped = currentIn.skip(bytesRemaining);
+            if (bytesSkipped == 0) {
+                int byteRead = read();
+                if (byteRead == -1) {
+                    break;
+                } else {
+                    bytesRemaining--;
+                }
+            } else {
+                bytesRemaining -= bytesSkipped;
+            }
+        }
+        return n - bytesRemaining;
+    }
+
+    @Override
+    public int available() throws IOException {
```

Comment:
I have added the javadoc explaining that the returned value for the available bytes is a conservative lower bound, because the subsequent components might also have available bytes but this method only returns the available bytes of the current component.